core/schema/middlewares.go

```diff
@@ -0,0 +1,137 @@
+package schema
+
+import (
+	"context"
+
+	"github.com/dagger/dagger/core"
+	"github.com/dagger/dagger/dagql"
+	"github.com/moby/buildkit/client/llb"
+)
+
+// DagOpMiddleware caches an arbitrary dagql field as a buildkit operation
+func DagOpMiddleware[T dagql.Typed, A any, R dagql.Typed](srv *dagql.Server, fn func(ctx context.Context, self dagql.Instance[T], args A) (R, error)) func(ctx context.Context, self dagql.Instance[T], args A) (R, error) {
+	return func(ctx context.Context, self dagql.Instance[T], args A) (inst R, err error) {
+		if _, ok := core.DagOpFromContext[core.RawDagOp](ctx); ok {
+			return fn(ctx, self, args)
+		}
+
+		deps, err := extractLLBDependencies(ctx, self.Self)
+		if err != nil {
+			return inst, err
+		}
+		return core.NewRawDagOp[R](ctx, srv, dagql.CurrentID(ctx).WithMetadata("", true), deps)
+	}
+}
+
+// DagOpFileMiddleware caches a file field as a buildkit operation - this is
+// more specialized than DagOpMiddleware, since that serializes the value to
+// JSON, so we'd just end up with a cached ID instead of the actual content.
+func DagOpFileMiddleware[T dagql.Typed, A any](srv *dagql.Server, fn func(ctx context.Context, self dagql.Instance[T], args A) (dagql.Instance[*core.File], error)) func(ctx context.Context, self dagql.Instance[T], args A) (dagql.Instance[*core.File], error) {
+	return func(ctx context.Context, self dagql.Instance[T], args A) (inst dagql.Instance[*core.File], err error) {
+		if _, ok := core.DagOpFromContext[core.DirectoryDagOp](ctx); ok {
+			return fn(ctx, self, args)
+		}
+
+		deps, err := extractLLBDependencies(ctx, self.Self)
+		if err != nil {
+			return inst, err
+		}
+
+		filename := "file"
+		id, err := srv.SelectID(ctx, srv.Root(),
+			dagql.Selector{
+				Field: "directory",
+			},
+			dagql.Selector{
+				Field: "withFile",
+				Args: []dagql.NamedInput{
+					{
+						Name:  "path",
+						Value: dagql.String(filename),
+					},
+					{
+						Name:  "source",
+						Value: dagql.NewID[*core.File](dagql.CurrentID(ctx).WithMetadata("", true)),
```

Comment:
Why do we set the digest to empty string for these? Not sure if that confuses the dagql cache and/or telemetry.

Reply:
Hah, the description of `WithMetadata` actually does cover this:

> `customDigest, if not empty string, will become the ID's digest.`

That said, yes, confusing. I've added a clearer shorthand `ID.WithTaint` to clarify the intention in these cases.

Reply:
Lol, forgetting things I wrote, 100% agree the more clear name is needed here 😄